README.TXT
MyEPG 0.2.5
Program released under the license GNU GPL v2
All credits are in CREDITS.TXT

** INSTALLATION

How to install it:
through the web interface go to Systems->Plugins
In install plugin select the file "myepg-(version)-dgs-sh4.tar.gz" and press "Upload plugin"

** GENERAL INFORMATION

The db files are three. By default they are:
- /media/myepg.headers.db
- /media/myepg.descriptors.db
- /media/myepg.aliases.db

The log files are three. By default thery are:
- /media/myepg.log
- /media/myepg.downloader.log
- /media/myepg.sync.log

The path can be changed from che configration file ('/var/plug-in/myepg.config') or from the configuration menu

Remote keys:
- Arrows		= grid navigation (hold pressed for a fast navigation)
- Ok			= change channel or set the zap (or remove it if already exist)
- Rec			= set the record (or remove it if already exist)
- Page up/down  = change the fav group
- Play			= go to the program transmitted now
- Vol. up/down  = channels page up/down
- i				= more info
- Forward		= go 24h forward
- Backward		= go 24h backward
- Exit 			= quit from plugin

** HINTS AND TIPS

There are some problems with original dgs epg db. It can contain only a limited number of events.
MyEPG download and manage a larger number of events so, if you put all the events into the internal dgs db,
many events are automatically deleted by the dgs software.

Actually the only solution is insert a limited number of events playing with 2 values (in configuration menu):
- sync hours	(how many hours sync with internal db)
- sync groups	(how groups sync with internal db)
They works as a filter on data sync.


ChangeLog
2009-08-05 [r32]  

	* (enigma2) fixed bug on language charset (common) add language iso
	  639 column in cvs format (enigma2) added italian language
	  translation (thanks to Spaeleus) (common) added iso 639 field in
	  epgdb (common) moved epgdb to revision 6 (enigma2) added ppc
	  makefiles (not working yet)

2009-07-26 [r31]  

	* (dgs) fixed background for light view (thanks to
	  ataru_moroboshi80) (dgs) fixed bad fifo management for
	  /dev/weboutput (dgs) fixed for new importer (with gzip
	  compression) (common) switched to version 0.3.1 (0.3.0 skipped
	  for ipk version compability)

2009-07-19 [r29-30]  

	* (common) remove useless files

	* (common) added gzip compression for csv (enigma2) added import
	  and scripts on daily action (common) fixed otv scanner (now
	  compile again)

2009-07-19 [r28]  

	* (enigma2) fixed bad hide handling for windows (enigma2) added
	  automatic script execution before import (for now only for manual
	  action) (enigma2) fixed bad epg path handling on edg nemesis
	  (untested) (enigma2) fixed loader with safe epg copy

2009-07-19 [r27]  

	* (common) added australiasat provider (common) added huffman debug

2009-07-16 [r26]  

	* (common) committed new providers configurations

2009-07-16 [r25]  

	* (common) delete old providers (common) delete useless directory

2009-07-16 [r24]  

	* (enigma2) added extra menu (enigma2) downloader could be setted
	  as plugin or extension in setup (enigma2) moved setup panel in
	  "main menu" -> "configuration" -> "crossepg" (enigma2) added
	  execution form (for importer script) (enigma2) other small fixes

2009-07-13 [r23]  

	* (enigma2) completed csv importer (enigma2) moved setup in enigma
	  configuration menu (enigma2) added configuration option
	  enabled/disable importer (enigma2) added configuration option
	  show/hide downloader in plugin list (enigma2) added skin files
	  for extra menu

2009-07-12 [r22]  

	* (enigma2) fixed again mjd calculation in importer (enigma2) added
	  interactive mode to importer (not yet finished)

2009-07-11 [r21]  

	* (enigma2) added aliases support in dbconverter

2009-07-11 [r20]  

	* (enigma2) added gettext support in python plugin

2009-07-11 [r19]  

	* (common) added svn2log for changelog (enigma2) fixed return value
	  on ipk preinst script

2009-07-10 [r18]  

	* (enigma2) fixed ipk for overwriting files

2009-07-10 [r17]  

	* (enigma2) changed ipkg builder script

2009-07-10 [r16]  

	* (enigma2) added ipkg packaging (enigma2) fixed python byte
	  compile (common) fixed a bug in mjd date in importer (enigma2)
	  fixed blank label in crossepg_setup

2009-07-09 [r15]  

	* (enigma2) added python compiler (enigma2) added ipkg packaging
	  (enigma2) fixed crossepg_dbinfo missing include

2009-07-08 [r13]  

	* (enigma2) added automatic enigma2 reboot (enigma2) added delay on
	  daily download if exist a record in progress (enigma2) added
	  crossepg_main.py for better integration with 3rd part panels
	  (enigma2) added crossepg_epgcopy for safe epg.dat copy

2009-07-06 [r12]  

	* (enigma2) added interactive do dbinfo (enigma2) added dump
	  database to dbinfo (enigma2) added info screen in python plugin

2009-07-05 [r11]  

	* (enigma2) added crossepg_dbinfo utility (enigma1) added some
	  makefiles (doesn't work yet)

2009-07-05 [r10]  

	* (enigma2) added autoreboot function for not patched images
	  (enigma2) fixed crossepg_epgmove.sh if configuration doesn't
	  exist (common) move epgdb to revision 5 (common) small changes in
	  epgdb (common) added creation date and update date in epgdb
	  (enigma2) fixed fault with custom lamedb in crossepg_dbconverter
	  (enigma2) re-added crossepg_importer

2009-07-04 [r9]  

	* (dgs) fixed scheduler problem with /dev/weboutput (dgs) fixed
	  crossepgd block disk suspend (dgs) added "show titles"
	  customization in aspect menu (dgs) little ui fixs (dgs) fixed bad
	  image size in preview box with HD channels

2009-07-03 [r7-8]  

	* (dgs) removed binary file from svn repository

	* (common) added missing files from last svn sync (enigma1) added
	  base structure and fixed common downloader

2009-07-03 [r6]  

	* (dgs) applied new epgdb (dgs) change light background (thanks to
	  ataru_moroboshi80) (common) updated package and clean script
	  (enigma2) fixed bad path generated by setup

2009-07-02 [r5]  

	* Fixed bug "cannot load epgdb" or "cannot read epgdb" at first
	  boot Published as 0.3.0rc1

2009-07-02 [r3-4]  

	* Removed python plugin duplicated files

	* Removed binaries

2009-07-02 [r2]  

	* First import



svn2log.py
#
# Copyright (c) 2003 The University of Wroclaw.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#    1. Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#    2. Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#    3. The name of the University may not be used to endorse or promote
#       products derived from this software without specific prior
#       written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE UNIVERSITY ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
# NO EVENT SHALL THE UNIVERSITY BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
# TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# History:
# 
# 2006-08-03  Przedsiebiorstwo Informatyczne CORE  <biuro at core.com.pl>:
#         * Following options were added:
#           -s, --strip-comments strip /* ... */ comments in log
#           -O, --only-date      generate only dates (without time)
#           -L, --no-files       generate log without filenames
#           -A, --no-author      generate log without author names
#           -H, --no-host        generate author name without hostname
#


import sys
import os
import time
import re
import getopt
import string
import codecs

from xml.utils import qp_xml

kill_prefix_rx = None
default_domain = "localhost"
exclude = []
users = { }
reloc = { }
max_join_delta = 3 * 60
list_format = False
strip = False
date_only = False
no_files = False
no_host = False
no_author = False

date_rx = re.compile(r"^(\d+-\d+-\d+T\d+:\d+:\d+)")

def die(msg):
  sys.stderr.write(msg + "\n")
  sys.exit(1)

def attr(e, n):
  return e.attrs[("", n)]

def has_child(e, n):
  for c in e.children:
    if c.name == n: return 1
  return 0

def child(e, n):
  for c in e.children:
    if c.name == n: return c
  die("<%s> doesn't have <%s> child" % (e.name, n))
  
def convert_path(n):
  for src in reloc.keys():
    n = string.replace(n, src, reloc[src])
  if kill_prefix_rx != None:
    if kill_prefix_rx.search(n):
      n = kill_prefix_rx.sub("", n)
    else:
      return None
  if n.startswith("/"): n = n[1:]
  if n == "": n = "/"
  for pref in exclude:
    if n.startswith(pref):
      return None
  return n

def convert_user(u):
  if no_author == False:
    if users.has_key(u):
      return users[u]
    else:
      if no_host:
        return u + ":"   
      else:
        return "%s <%s@%s>:" % (u, u, default_domain)
  else:
    return ''

def wrap_text_line(str, pref, width):
  ret = u""
  line = u""
  first_line = True
  for word in str.split():
    if line == u"":
      line = word
    else:
      if len(line + u" " + word) > width:
        if first_line:
          ret += line + u"\n"
          first_line = False
          line = word
        else:
          ret += pref + line + u"\n"
          line = word
      else:
        line += u" " + word
  if first_line:
    ret += line + u"\n"
  else:
    ret += pref + line + u"\n"
  return ret

def wrap_text(str, pref, width):
  if not list_format:
    return wrap_text_line(str,pref,width)
  else:
    items = re.split(r"\-\s+",str)
    ret = wrap_text_line(items[0],pref,width)
    for item in items[1:]:
      ret += pref + u"- " + wrap_text_line(item,pref+"  ",width)
    return ret

class Entry:
  def __init__(self, tm, rev, author, msg):
    self.tm = tm
    self.rev = rev
    self.author = author
    self.msg = msg
    self.beg_tm = tm
    self.beg_rev = rev

  def join(self, other):
    self.tm = other.tm
    self.rev = other.rev
    self.msg += other.msg

  def dump(self, out):
    if len(self.msg) > 0:
      if date_only == False:
        tformat = "%Y-%m-%d %H:%M +0000"
      else:
        tformat = "%Y-%m-%d"

      if self.rev != self.beg_rev:
        out.write("%s [r%s-%s]  %s\n\n" % \
                          (time.strftime(tformat, time.localtime(self.beg_tm)), \
                           self.rev, self.beg_rev, convert_user(self.author)))
      else:
        out.write("%s [r%s]  %s\n\n" % \
                          (time.strftime(tformat, time.localtime(self.beg_tm)), \
                           self.rev, convert_user(self.author)))
      out.write(self.msg)
  
  def can_join(self, other):
    return self.author == other.author and abs(self.tm - other.tm) < max_join_delta

def process_entry(e):
  rev = attr(e, "revision")
  if has_child(e, "author"):
    author = child(e, "author").textof()
  else:
    author = "anonymous"
  m = date_rx.search(child(e, "date").textof())
  msg = ' ' + child(e, "msg").textof()
  if strip == True:
    ibegin = string.find(msg, "/*")
    if ibegin > 0:
      iend = string.find(msg, "*/") + 2
      msg = msg[0:ibegin] + msg[iend:]

  if m:
    tm = time.mktime(time.strptime(m.group(1), "%Y-%m-%dT%H:%M:%S"))
  else:
    die("evil date: %s" % child(e, "date").textof())
  paths = []
  if len(msg) > 1: 
    for path in child(e, "paths").children:
      if path.name != "path": die("<paths> has non-<path> child")
      nam = convert_path(path.textof())
      if nam != None:
        if attr(path, "action") == "D":
          paths.append(nam + " (removed)")
        elif attr(path, "action") == "A":
          paths.append(nam + " (added)")
        else:
          paths.append(nam)
     
    if paths != [] and no_files == False:
      return Entry(tm, rev, author, "\t* %s\n" % wrap_text(", ".join(paths) + ": " + msg, "\t  ", 65))
    elif paths != [] and no_files == True:
      return Entry(tm, rev, author, "\t* %s\n" % wrap_text(msg, "\t  ", 65))

  return None

def process(fin, fout):
  parser = qp_xml.Parser()
  root = parser.parse(fin)

  if root.name != "log": die("root is not <log>")
  
  cur = None
  
  for logentry in root.children:
    if logentry.name != "logentry": die("non <logentry> <log> child")
    e = process_entry(logentry)
    if e != None:
      if cur != None:
        if cur.can_join(e):
          cur.join(e)
        else:
          cur.dump(fout)
          cur = e
      else: cur = e
        
  if cur != None: cur.dump(fout)

def usage():
  sys.stderr.write(\
"""Usage: %s [OPTIONS] [FILE]
Convert specified subversion xml logfile to GNU-style ChangeLog.

Options:
  -p, --prefix=REGEXP  set root directory of project (it will be striped off
                       from ChangeLog entries, paths outside it will be 
                       ignored)
  -x, --exclude=DIR    exclude DIR from ChangeLog (relative to prefix)
  -o, --output         set output file (defaults to 'ChangeLog')
  -d, --domain=DOMAIN  set default domain for logins not listed in users file
  -u, --users=FILE     read logins from specified file
  -F, --list-format    format commit logs with enumerated change list (items
                       prefixed by '- ')
  -r, --relocate=X=Y   before doing any other operations on paths, replace
                       X with Y (useful for directory moves)
  -D, --delta=SECS     when log entries differ by less then SECS seconds and
                       have the same author -- they are merged, it defaults
                       to 180 seconds
  -h, --help           print this information
  -s, --strip-comments strip /* ... */ comments in log
  -O, --only-date      generate only dates (without time)
  -L, --no-files       generate log without filenames
  -A, --no-author      generate log without author names
  -H, --no-host        generate author name without hostname

Users file is used to map svn logins to real names to appear in ChangeLog.
If login is not found in users file "login <login@domain>" is used.

Example users file:
john    John X. Foo <jfoo@example.org>
mark    Marcus Blah <mb@example.org>

Typical usage of this script is something like this:

  svn log -v --xml | %s -p '/foo/(branches/[^/]+|trunk)' -u aux/users
  
Please send bug reports and comments to author:
  Michal Moskal <malekith@pld-linux.org>
  
Regarding -s, -O, -L, -A, -H options see
  http://www.core.com.pl/svn2log
   

""" % (sys.argv[0], sys.argv[0]))

def utf_open(name, mode):
  return codecs.open(name, mode, encoding="utf-8", errors="replace")

def process_opts():
  try:
    opts, args = getopt.gnu_getopt(sys.argv[1:], "o:u:p:x:d:r:d:D:FhsOLHA", 
                                   ["users=", "prefix=", "domain=", "delta=",
                                    "exclude=", "help", "output=", "relocate=",
                                    "list-format","strip-comments", "only-date", "no-files",
																		"no-host", "no-author"])
  except getopt.GetoptError:
    usage()
    sys.exit(2)
  fin = sys.stdin
  fout = None
  global kill_prefix_rx, exclude, users, default_domain, reloc, max_join_delta, list_format, strip, date_only, no_files, no_host, no_author
  for o, a in opts:
    if o in ("--prefix", "-p"):
      kill_prefix_rx = re.compile("^" + a)
    elif o in ("--exclude", "-x"):
      exclude.append(a)
    elif o in ("--help", "-h"):
      usage()
      sys.exit(0)
    elif o in ("--output", "-o"):
      fout = utf_open(a, "w")
    elif o in ("--domain", "-d"):
      default_domain = a
    elif o in ("--strip-comments", "-s"):
      strip = True
    elif o in ("--only-date", "-O"):
      date_only = True
    elif o in ("--no-files", "-L"):
      no_files = True
    elif o in ("--no-host", "-H"):
      no_host = True
    elif o in ("--no-author", "-A"):
      no_author = True
    elif o in ("--users", "-u"):
      f = utf_open(a, "r")
      for line in f.xreadlines():
        w = line.split()
        if len(line) < 1 or line[0] == '#' or len(w) < 2: 
          continue
        users[w[0]] = " ".join(w[1:])
    elif o in ("--relocate", "-r"):
      (src, target) = a.split("=")
      reloc[src] = target
    elif o in ("--delta", "-D"):
      max_join_delta = int(a)
    elif o in ("--list-format", "-F"):
      list_format = True
    else:
      usage()
      sys.exit(2)
  if len(args) > 1:
    usage()
    sys.exit(2)
  if len(args) == 1:
    fin = open(args[0], "r")
  if fout == None:
    fout = utf_open("ChangeLog", "w")
  process(fin, fout)

if __name__ == "__main__":
  os.environ['TZ'] = 'UTC'
  try:
    time.tzset()
  except AttributeError:
    pass
  process_opts()